### PR TITLE
Add typings for the promise-hash library, v1.3

### PR DESCRIPTION
**Pull request by mshafir**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Adding a new definition:
- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [X] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [X] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [X] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [X] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [X] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/promise-hash/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+declare global {
+    interface PromiseConstructor {
+        hash<T>(promiseHash: { [P in keyof T]: PromiseLike<T[P]> | T[P] }):
+            Promise<{ [P in keyof T]: T[P] }>;
+    }
+}
+
+declare function hash<T>(
+    promiseHash: { [P in keyof T]: PromiseLike<T[P]> | T[P]}):
+    Promise<{ [P in keyof T]: T[P] }>;
```
Comment (PranavSenthilnathan) [inline types/promise-hash/index.d.ts]:
```suggestion
    Promise<T>;
```

File: types/promise-hash/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Type definitions for promise-hash 1.3
+// Project: https://github.com/mtimofiiv/promise-hash
+// Definitions by: Michael Shafir <https://github.com/mshafir>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export = hash;
+
+declare global {
+    interface PromiseConstructor {
+        hash<T>(promiseHash: { [P in keyof T]: PromiseLike<T[P]> | T[P] }):
```
Comment (PranavSenthilnathan) [inline types/promise-hash/index.d.ts]: You can store the function type and reuse it here as well as for the declared function below.

File: types/promise-hash/index.d.ts (hunk 1 above)
Comment (mshafir) [reply types/promise-hash/index.d.ts]: Making this change makes the tests fail, though it's not entirely clear to me why.
`ERROR: 12:1  expect  TypeScript@next expected type to be:
  Promise<{ a: number; b: string; c: { d: boolean; e: string; }; }>
got:
  Promise<{ a: number; b: string; c: { d: any; e: any; }; }>`

File: types/promise-hash/index.d.ts (hunk 2 above)
Comment (mshafir) [reply types/promise-hash/index.d.ts]: I'm not sure how best to do that. The most I could figure out is a shared type for the arguments with
```ts
type PromiseHash<T> = { [P in keyof T]: PromiseLike<T[P]> | T[P] };
```
I couldn't figure out how to get the whole function type to be reused.

File: types/promise-hash/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// Type definitions for promise-hash 1.3
+// Project: https://github.com/mtimofiiv/promise-hash
+// Definitions by: Michael Shafir <https://github.com/mshafir>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export = hash;
+
+type PromiseHash<T> = { [P in keyof T]: PromiseLike<T[P]> | T[P] };
```
Comment (PranavSenthilnathan) [inline types/promise-hash/index.d.ts]: I was thinking along the lines of [this](https://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=22&pc=57#code/C4TwDgpgBACgTgewLYEsDOEASBDNALKAXigB4AVAPgAoxFUMd8AuKAbygG0YoUA7KANYQQCAGZQyAXRbxk6CABkUQ8l0kUoAHwlqoAXwCURDbPoRyFANwAoawBMIAYwA22ONADmzhACNsztmsoYJ5eYAg4UWxHaFN5AGEEXjRgOABXR2AEOECQvKg8XDwZOnlGPBs8vWtq+ydXdyhHJJSCopK5BiKba2bk4Ch3NDTnAeJC-CpWIJDsDrMAOiGEZwA3CCoAVgMAGhngn3n5JYg0FfWqACJwlMvd-aaWCbwph7y7I4wTs7WN1LSIPd8iEIJ8IN9zhtLtgPNg+HcHoYagYeqBIFAxrBShgSOw5lBeGkkD4IpYoIcoCk4HwPGTHCx2B9yQgVhBsLwyaDKakaWS9HyrNY0dAAB5EDHgCBiQanEYDCAi8K8OxoDFQAD8GPS0BYUWcGDJAHpDVqAULJVAQOL5YqIMrVcLpUM5RrTTqoHqDVBjW6gA). I think it's just a matter of preference, so I'm happy to merge either one.

Comment (mshafir) [reply types/promise-hash/index.d.ts]: Cool! I had no idea you could put the generic <T> in front of the anonymous function like that. I'll make the update.

**Comment by PranavSenthilnathan** [comment]
Congratulations on your first DefinitelyTyped contribution! Thank you for being a part of the community!

**Pull request merged**